File: src/Algorithm/Searching/Leetcode_problems/Search_in_rotated_sorted_array/Search_in_rotated_sorted_array.cpp
class Solution {
public:
    int binarysearch(vector<int>& nums,int target,int first,int last)
    {
        while(first<=last)
        {
            int midval=(first+last)/2;
            if(nums[midval]==target) return midval;
            else if(nums[midval]>target)
            {
                last=midval-1;
            }
            else {first=midval+1;}
        }
        
        return -1;
    }
    int search(vector<int>& nums, int target) {
        int n=nums.size();
        int start=0,end=n-1,mid=-1,index=-1;
        while(start<=end)
        {
            mid=(start+end)/2;
            if(nums[start]<=nums[end]) {index=start;break;}
            else if(nums[mid]>nums[end]) start=mid+1;
            else end=mid;
        }
        int num1= binarysearch(nums,target,0,index-1);
        int num2= binarysearch(nums,target,index,n-1);
        if(num1==-1)return num2;
        else return num1;
    }
};

File: src/Striver_SDE_Sheet/Binary_Search/Search_in_Rotated_Sorted_Array/SearchinRotatedSortedArray.java
package Striver_SDE_Sheet.Binary_Search.Search_in_Rotated_Sorted_Array;

public class SearchinRotatedSortedArray {
    public static void main(String[] args){
        int[] nums = {4,5,6,7,0,1,2};
        int target = 0;
        System.out.println(search(nums,target));
    }

    private static int search(int[] nums, int target) {

        int start = 0;
        int end = nums.length - 1;
        while (start < end) {
            int mid = (start + end) / 2;

            if (nums[mid] == target) return mid;

            if (nums[start] <= nums[mid]) {
                if (target >= nums[start] && target < nums[mid]) {
                    end = mid - 1;
                } else {
                    start = mid + 1;
                }
            } else {
                if (target > nums[mid] && target <= nums[end]) {
                    start = mid + 1;
                } else {
                    end = mid - 1;
                }
            }
        }
        return nums[start] == target ? start : -1;
    }
}
